src/Analyser/NodeScopeResolver.php

```diff
@@ -791,6 +793,10 @@ private function processStmtNode(
 					$classReflection,
 					$methodReflection,
 				), $methodScope);
+
+				if ($isConstructor && $this->narrowMethodScopeFromConstructor) {
+					$scope = $statementResult->getScope()->rememberConstructorScope();
```

Comment:
I don't think `$statementResult->getScope()` contains what we want here. Consider branches that always end in `throw` - the object isn't actually created for that situation.

Please add a test for this scenario and check what happens. I think what we need is something like this: https://github.com/phpstan/phpstan-src/blob/375f68eb71490589ec0b1fa810ac190761ef7ba3/src/Rules/TooWideTypehints/TooWideParameterOutTypeCheck.php#L34-L51

Getting the types from execution ends and return statements is something we need I think here: https://github.com/phpstan/phpstan-src/blob/375f68eb71490589ec0b1fa810ac190761ef7ba3/src/Rules/TooWideTypehints/TooWideParameterOutTypeCheck.php#L34-L51